eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Alonzo.Binary.CostModelsSpec (spec) where
+
+import Cardano.Ledger.Alonzo.Core
+import Cardano.Ledger.Alonzo.Scripts
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Binary (ToCBOR (..))
+import Cardano.Ledger.Binary.Decoding
+import Cardano.Ledger.Binary.Encoding
+import Cardano.Ledger.Plutus.Language
+import qualified Data.Map as Map
+import Data.Text (Text)
+import Data.Typeable
+import Data.Word (Word8)
+import Lens.Micro
+import Test.Cardano.Ledger.Alonzo.Arbitrary
+import Test.Cardano.Ledger.Common
+
+spec ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+spec pxy = do
+  describe "CostModels CBOR deserialization" $ do
+    validCostModelProp @era pxy
+    invalidCostModelProp @era Proxy
+    unknownCostModelProp @era Proxy
+
+validCostModelProp ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+validCostModelProp pxy = do
+  prop "valid CostModels deserialize correctly, both independently and within PParamsUpdate" $
+    \(lang :: Language) -> do
+      forAllShow (genValidCostModelsEnc lang) (showEnc pxy) $
+        \validCmEnc -> do
+          encodeAndCheckDecoded @era validCmEnc $
+            \cmDecoded ppuDecoded -> do
+              expectRight cmDecoded >>= (`shouldSatisfy` validCm)
+              ppuRes <- expectRight ppuDecoded
+              ppuRes `shouldSatisfy` \ppu -> (validCm <$> ppu ^. ppuCostModelsL) == SJust True
+  where
+    genValidCostModelsEnc lang = genCostModelsEncForLanguage lang (costModelParamsCount lang)
```

Comment:
Minor nit pick. Each language has one cost model, so it is not plural. The only reason why I am suggesting it, is because it can be confused with generator that produces CostModels for many languages (which even has a separate type `CostModels`) 

```suggestion
    genValidCostModelEnc lang = genCostModelEncForLanguage lang (costModelParamsCount lang)
```

Reply:
Yes completely agree.